homeassistant/components/izone/update_coordinator.py

```diff
@@ -0,0 +1,174 @@
+    """
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        entry: ConfigEntry,
+        ctrl: pizone.Controller,
+    ) -> None:
+        """Initialise the update coordinator."""
+        self._controller = ctrl
+        self._available_listeners: list[CALLBACK_TYPE] = []
+        self._available = True
+
+        @callback
+        def controller_disconnected(ctrl: pizone.Controller, _: Exception) -> None:
+            """Disconnected from controller."""
+            _LOGGER.info(
+                "Controller '%s' disconnected",
+                self._controller.device_uid,
+                exc_info=True,
+            )
+            if ctrl is not self._controller:
+                return
+            self._set_available(False)
+
+        entry.async_on_unload(
+            async_dispatcher_connect(
+                hass, DISPATCH_CONTROLLER_DISCONNECTED, controller_disconnected
+            )
+        )
+
+        @callback
+        def controller_reconnected(ctrl: pizone.Controller) -> None:
+            """Reconnected to controller."""
+            _LOGGER.info(
+                "Controller '%s' reconnected",
+                self._controller.device_uid,
+                exc_info=True,
+            )
+            if ctrl is not self._controller:
+                return
+            self._set_available(True)
+
+        entry.async_on_unload(
+            async_dispatcher_connect(
+                hass, DISPATCH_CONTROLLER_RECONNECTED, controller_reconnected
+            )
+        )
+
+    def _set_available(self, available: bool):
+        self._available = available
+        for listener in self._available_listeners:
+            listener()
+
+    @callback
+    def async_add_available_listener(
+        self, update_callback: CALLBACK_TYPE
+    ) -> Callable[[], None]:
+        """Listen for data updates."""
+        self._available_listeners.append(update_callback)
+
+        @callback
+        def remove_listener() -> None:
+            """Remove update listener."""
+            self.async_remove_available_listener(update_callback)
+
+        return remove_listener
+
+    @callback
+    def async_remove_available_listener(self, update_callback: CALLBACK_TYPE) -> None:
+        """Remove data update."""
+        self._available_listeners.remove(update_callback)
+
+    @property
+    def controller(self) -> pizone.Controller:
+        """Return the pizone controller from the API."""
+        return self._controller
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        return self._available
+
+    async def async_request_refresh(self) -> None:
+        """Request a refresh. This gets passed off to the controller and batched."""
+        await self._controller.refresh()
+
+
+ControllerAddedListener = Callable[[ControllerUpdateCoordinator], None]
+
+
+class ControllerCoordinatorEntity(Entity):
```

Comment:
The base entity we generally put in a separate module (`entity.py`), please consider moving it.

Reply:
`homeassistant\helpers\update_coordinator.py` contains both `DataUpdateCoordinator` and `CoordinatorEntity`. I was trying to keep with that. Can move if you like though.